progresso/__init__.py

```diff
@@ -0,0 +1,219 @@
+                           + self.invisible_chars)
+        sys.stdout.write(self.format(line))
+
+        if self.fraction >= 1.0:
+            self.finish()
+        else:
+            sys.stdout.flush()
+
+    def finish(self):
+        sys.stdout.write(self.format(self.outro))
+
+    def __call__(self, **kwargs):
+        self.update(**kwargs)
+
+    def update(self, step=1, **kwargs):
+        if step:
+            self.done += step
+        # Allow to override any properties.
+        self.__dict__.update(kwargs)
+        if self.start is None and 'done' in kwargs:
+            # First call to update and forcing a done value. May be
+            # resuming a download. Keep track for better ETA computation.
+            self.supply = self.done
+        self.render()
+
+    def __next__(self):
+        self.update()
+
+    def iter(self, iterable):
+        for i in iterable:
+            yield i
+            self.update()
+        if self.fraction != 1.0:
+            # Spinner without total.
+            self.finish()
+
+
+# Manage sane default formats while keeping the original type to allow any
+# built-in formatting syntax.
+
+class Float(float):
+
+    def __format__(self, format_spec):
+        if not format_spec:
+            format_spec = '.2f'
+        return super().__format__(format_spec)
+
+
+class Percent(float):
+
+    def __format__(self, format_spec):
+        if not format_spec:
+            format_spec = '.2%'
+        return super().__format__(format_spec)
+
+
+class ETA(datetime):
+
+    def __format__(self, format_spec):
+        if not format_spec:
+            now = datetime.now()
+            diff = self - now
+            format_spec = '%H:%M:%S'
+            if diff.days > 1:
+                format_spec = '%Y-%m-%d %H:%M:%S'
+        return super().__format__(format_spec)
+
+    @classmethod
+    def from_datetime(cls, dt):
+        # Find a more elegant way.
+        return cls(year=dt.year, month=dt.month, day=dt.day, hour=dt.hour,
+                   minute=dt.minute, second=dt.second, tzinfo=dt.tzinfo)
+
+
+class Timedelta(int):
+    """An integer that is formatted by default as timedelta."""
+
+    def format_as_timedelta(self):
+        """Format seconds as timedelta."""
+        # Do we want this as a Formatter type also?
+        tmp = timedelta(seconds=self)
+        # Filter out microseconds from str format.
+        # timedelta does not have a __format__ method, and we don't want to
+        # recode it (we would need to handle i18n of "days").
+        obj = timedelta(days=tmp.days, seconds=tmp.seconds)
+        return str(obj)
+
+    def __format__(self, format_spec):
+        if not format_spec:
+            return self.format_as_timedelta()
+        return super().__format__(format_spec)
+
+
+VERSION = (0, 0, 1)
+
+__author__ = 'Yohan Boniface'
+__contact__ = "hi@yohanboniface.me"
+__homepage__ = "https://github.com/yohanboniface/progresso"
+__version__ = ".".join(map(str, VERSION))
```

Comment:
If you are referring to metadatas being in `__init__.py`, I'd say that this allows to do `import pkgname; print(pkgname.__whatever__`
If you are referring to version string built from a tuple, I'd say because distuils wants a string, but we prefer keep a structured data internally.
That said, I must confess that the way I'm doing packaging is copypastasciutta from a project to the other, trying to make it better on the go. So I'm happy to learn best practices on this :)